File: src/Practice/Sorting/MergeSort.java
package Sorting;

import Datastructures.bk;

/**
 * Created by BK on 11-09-2017.
 */
public class MergeSort {
    static MergeTwoSortedArrays merge;

    public static void main(String[] args) {
        merge = new MergeTwoSortedArrays();
        int[] input = {4, 2, 12, 3, 13, 7, 1};
        bk.print_int(input);
        int[] sorted = mergeSort(input);
        bk.print_int(sorted);
    }

    public static int[] mergeSort(int[] input) {
        int[] sorted = new int[0];
        int[] left = leftArray(input, input.length / 2);
        int[] right = rightArray(input, left.length);
        if (left.length!=1){
            mergeSort(left);
        } else {
            System.out.print("Left : ");
            bk.print_int(left);
            System.out.print("Right : ");
            bk.print_int(right);
        }
        if (right.length!=1){
            mergeSort(right);
        }
        return sorted;
    }

    private static int[] rightArray(int[] input, int from) {
        int[] temp = new int[input.length - from];
        for (int i = 0; i < temp.length; i++) temp[i] = input[from + i];
//        System.out.print("Right : ");
//        bk.print_int(temp);
        return temp;
    }

    private static int[] leftArray(int[] input, int len) {
        int[] temp = new int[len];
        for (int i = 0; i < temp.length; i++) temp[i] = input[i];
//        System.out.print("Left : ");
//        bk.print_int(temp);
        return temp;
    }
}

File: src/Practice/Sorting/merge-sort.py
def merge(arr, left, right):
    i = 0
    j = 0
    k = 0
    print("arr is ", arr, " left is ", left, " right is ", right)
    while i < len(left) and j < len(right):
        if left[i] > right[j]:
            arr[k] = left[i]
            i += 1
        else:
            arr[k] = right[j]
            j += 1
        k += 1

    while i < len(left):
        arr[k] = left[i]
        k += 1
        i += 1

    while j < len(right):
        arr[k] = right[j]
        k += 1
        j += 1
    print("merged is ",arr)


def merge_sort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        left = arr[:mid]
        right = arr[mid:]
        merge_sort(left)
        merge_sort(right)

        """
        At each stage we take the array and split into left half and right half
        
        At that stage itself we merge it by the principle of merging two sorted halves
        
        So by end of that recursion stage the half ( either left or right ) becomes sorted
        
        """

        merge(arr, left, right)


def main():
    unsorted = [3, 463, 435, 46, 39, 9, 2, 1]
    print(unsorted)
    merge_sort(unsorted)
    print(unsorted)


main()
